### DISCUSSION TITLE
issue: Form does not trigger onChange when registering nested optional objects

### DISCUSSION THREAD

**Discussion by slhck** [react-hook-form/react-hook-form#10813] [Q&A] [answered]
### Version Number

7.45.2

### Codesandbox/Expo snack

https://codesandbox.io/s/practical-buck-cg5wt2?file=/src/App.tsx

### Steps to reproduce

1. Go to https://codesandbox.io/s/practical-buck-cg5wt2?file=/src/App.tsx
2. Click on the input checkbox

### Expected behaviour

I expect the form to call `onChange` and subsequently trigger the `useEffect` that enables the submit button.

When I comment out this:

```
{...register("credentials.foo.username")}
```

The form behaves as expected.

However, once that field is present, despite it being optional, the form does not allow submitting anymore.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
is this the same issue: https://github.com/react-hook-form/react-hook-form/issues/10744

**Comment by slhck** [comment]
I see a patch has been merged, thanks for pointing out the duplicate.

**Comment by slhck** [comment]
Note that this also does not work in 7.45.0, so it is not a direct duplicate of the issue.

I tried v7.45.3 and it still does not work:

[![Edit modern-bird-4rc6t8](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/modern-bird-4rc6t8?fontsize=14&hidenavigation=1&theme=dark)

Note that clicking the checkbox does not create a change in the form.

**Comment by slhck** [comment]
@bluebill1049 Any idea what this issue might be in relation to the other one? (It's not a duplicate but seems similar in behavior.)

**Comment by bluebill1049** [comment]
Does the latest fix this issue?

**Comment by leapful** [comment]
@slhck 

You only marked `credentials.foo` as optional field but then `crendetials.foo.username` is in email pattern therefore it's always invalid and prevent submit form.

Furthermore, you are using `useEffect` with change dependencies on `isDirty`, `isValid`, `errors` which are unchanged after clicking on checkbox field, reason: `isDirty` is always true since `credentials.foo.username` is now having empty string value due to bind with native HTML input element and different with your default value, `isValid` is always false due to your zod schema validation.

If you want to trigger any action on change of the field, just put it in your custom onChange handler.

**Reply by slhck** [reply]
Thanks for your answer!

I already updated to the latest version 7.45.4 now.

> You only marked credentials.foo as optional field but then crendetials.foo.username is in email pattern therefore it's always invalid and prevent submit form.

Even when I set it to `.optional()`, clicking the checkbox does not enable the submit button:

https://codesandbox.io/s/practical-buck-cg5wt2?file=/src/App.tsx

Also, the `errors` I get were always empty, not indicating that this part is missing.

> credentials.foo.username is now having empty string value due to bind with native HTML input element and different with your default value

How would I fix this?

**Reply by leapful** [reply]
@slhck 

>Even when I set it to .optional(), clicking the checkbox does not enable the submit button

=> zod `optional()` does not accept empty string as valid optional value, it always triggers email validation and throw error on empty string value.

>Also, the errors I get were always empty, not indicating that this part is missing. How would I fix this?

=> There is nothing wrong to fix, like I already explained in the previous comment, your change dependencies `[isDirty, isValid, errors]` for `useEffect` does not change over time in toggling your checkbox, therefore your handling inside `useEffect` only executes once.

You need to use `useWatch` on your checkbox field name to watch on change on its value or execute your handling directly inside `onChange` handler of checkbox field.

Working demo: https://codesandbox.io/s/friendly-jasper-l9z2hv?file=/src/App.tsx

**Reply by slhck** [reply, marked as answer]
Thanks for providing the demo. The button still was not enabled after clicking the checkbox, but I resolved this thanks to your pointer with:

```
username: z.coerce.string().email().optional().or(z.literal(""))
```

The underlying issue apparently is the email validation, for more context see: https://github.com/colinhacks/zod/issues/310
